CMakeLists.txt

```diff
@@ -0,0 +1,101 @@
+set(FINUFFT_ENABLE_SANITIZERS "ON" CACHE BOOL "Whether to enable sanitizers, only effective for Debug configuration.")
+
+if(FINUFFT_BUILD_MATLAB)
+    # When building for matlab, we will fetch the OpenMP library used by matlab
+    # instead of system default for compatibility.
+    find_package(Matlab REQUIRED)
+    find_library(OpenMP_iomp5_LIBRARY NAMES iomp5 HINTS ${Matlab_ROOT_DIR}/sys/os/ PATH_SUFFIXES glnxa64 maci64)
+    find_library(OpenMP_pthreads_LIBRARY NAMES pthread CMAKE_FIND_ROOT_PATH_BOTH)
+endif()
+
+# Find required packages
+list(APPEND CMAKE_MODULE_PATH "${CMAKE_CURRENT_SOURCE_DIR}/cmake")
+find_package(FFTW REQUIRED)
+find_package(OpenMP REQUIRED)
+
+set(FINUFFT_FFTW_LIBRARIES "FFTW::Float" "FFTW::Double" "FFTW::Float${FINUFFT_FFTW_SUFFIX}" "FFTW::Double${FINUFFT_FFTW_SUFFIX}")
+
+# This set of sources is compiled twice, once in single precision and once in double precision
+# The single precision compilation is done with -DSINGLE
+set(FINUFFT_PRECISION_DEPENDENT_SOURCES src/finufft.cpp src/simpleinterfaces.cpp src/spreadinterp.cpp src/utils.cpp fortran/finufftfort.cpp)
+
+# Utility function to enable ASAN on debug builds
+set(FINUFFT_ASAN_SUPPORTED_COMPILERS AppleClang Clang GNU)
+function(enable_asan target)
+    if(NOT FINUFFT_ENABLE_SANITIZERS)
+        return()
+    endif()
+
+    if(CMAKE_CXX_COMPILER_ID IN_LIST FINUFFT_ASAN_SUPPORTED_COMPILERS)
+        # Enable only on clang / gcc compilers.
+        target_compile_options(${target} PUBLIC $<$<CONFIG:DEBUG>:-fsanitize=address>)
+        target_link_options(${target} PUBLIC $<$<CONFIG:DEBUG>:-fsanitize=address>)
+    endif()
+endfunction()
+
+# Utility function to set finufft compilation options.
+function(set_finufft_options target)
+    set_property(TARGET ${target} PROPERTY POSITION_INDEPENDENT_CODE ON)
+    set_property(TARGET ${target} PROPERTY CMAKE_CXX_STANDARD 14)
+    enable_asan(${target})
+    target_compile_options(${target} PRIVATE SHELL:$<$<CONFIG:Release,RelWithDebInfo>:${FINUFFT_ARCH_FLAGS}>)
+    target_include_directories(${target} PUBLIC "${CMAKE_CURRENT_SOURCE_DIR}/include")
+    target_link_libraries(${target} PUBLIC OpenMP::OpenMP_CXX)
+endfunction()
+
+# Main finufft libraries
+add_library(finufft_f32 OBJECT ${FINUFFT_PRECISION_DEPENDENT_SOURCES})
+target_compile_definitions(finufft_f32 PRIVATE SINGLE)
+set_finufft_options(finufft_f32)
+target_link_libraries(finufft_f32 PUBLIC ${FINUFFT_FFTW_LIBRARIES})
+
+add_library(finufft_f64 OBJECT ${FINUFFT_PRECISION_DEPENDENT_SOURCES})
+set_finufft_options(finufft_f64)
+target_link_libraries(finufft_f64 PUBLIC ${FINUFFT_FFTW_LIBRARIES})
+
+add_library(finufft SHARED src/utils_precindep.cpp contrib/legendre_rule_fast.c)
+set_finufft_options(finufft)
+target_link_libraries(finufft PUBLIC finufft_f32 finufft_f64 m)
+target_include_directories(finufft PUBLIC "${CMAKE_CURRENT_SOURCE_DIR}/include")
+
+add_library(finufft_static STATIC src/utils_precindep.cpp contrib/legendre_rule_fast.c)
```

Comment:
All the .cpp files and /contrib/legendre_rule_fast.c can be compiled in objects inside finufft64.

I would do:
```cmake
add_library(finufft64 OBJECT ${SOURCES}) // SOURCES contains all .cpp and legendre_rule_fast.c
set_target_properties(finufft32 PROPERTIES POSITION_INDEPENDENT_CODE ON)
add_library(finufft_static STATIC $<TARGET_OBJECTS:finufft32> $<TARGET_OBJECTS:finufft64>)
target_link_libraries(finufft_static PRIVATE ${LIBRARIES})
target_include_directories(finufft_static PUBLIC ${CMAKE_CURRENT_SOURCE_DIR}/include) // this exports the header when they link against it

add_library(finufft_shared SHARED $<TARGET_OBJECTS:finufft32> $<TARGET_OBJECTS:finufft64>)
target_link_libraries(finufft_shared PRIVATE ${LIBRARIES}) //libraries contains fftw and openmp (if enabled)
target_include_directories(finufft_shared PUBLIC ${CMAKE_CURRENT_SOURCE_DIR}/include) // this exports the header when they link against it
```